ADC STM32: Various fixes and enhancements

gautierg-st opened the pull request:
This PR contains various fixes and enhancements for the STM32 ADC driver:
- Add a log message when an overrun error occurs (with indications how to prevent the error)
- Simplify the oversampling function by adding a new property in device tree
- Simplify the DMA enable function and add support for DMA for STM32H7A/H7B and F1 series

Fixes #79909

gautierg-st committed:
drivers: adc: stm32: add log message when adc overrun

Displays an error log message when an ADC overrun occurs.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: stm32: better express f3 and h7 adc versions

STM32F3 and H7 have multiple ADC versions difficult to differentiate.
Use clearer macros to make code more readable.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
dts: bindings: adc: stm32: add oversampler

On STM32 ADC, there are currently two types of oversamplers (if present),
one with 8 available oversampling values, the other with 1024.
To simplify the driver, add the oversampler as a dts property.
Also add defines to avoid magic values in the dtsi.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
dts: arm: st: add adc oversampler to all stm32 series

Add ADC oversampler type to all ADC of all series.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: stm32: simplify oversampling with new property

Use the new oversampler property to simplify the management of the
ADC oversampling.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: stm32: simplify adc dma enable function

Remove specific cases for H7 and U5: group them together and only call a
single function. ADC3 of H72x/H73x and ADC4 of U5 are different from other
ADC of their series, and have dedicated functions in the LL for enabling
DMA, but they're doing the exact same operation as
LL_ADC_REG_SetDataTransferMode.

Incidentally, this change allows H7A/H7B to use the DMA (it seems to have
been missed before).

Last, this change enables the DMA support for F1x ADC.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc: api: add test for nucleo_f103rb with dma

Adds an ADC DMA test for Nucleo F103RB board now that support for it is
added in the ADC driver.
Reading multiple channels in F1 is only possible with the DMA, hence the
necessity to have a dedicated twister test for it.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -230,38 +230,16 @@ struct adc_stm32_cfg {
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32f1_adc)
+	LL_ADC_REG_SetDMATransfer(adc, LL_ADC_REG_DMA_TRANSFER_UNLIMITED);
+#elif defined(CONFIG_SOC_SERIES_STM32H7X) || defined(CONFIG_SOC_SERIES_STM32U5X)
+	/* H72x ADC3 and U5 ADC4 are different from the rest, but this call works also for them,
+	 * so no need to call their specific function
+	 */
```
Open: Does commenting LL shenanigans adds clarity to the reader here ?

gautierg-st replied:
For a reader who would be aware that these ADCs are different (and I include myself in it), it mentions the fact that this is done on purpose, that the peculiarities of these ADC were taken into account.

erwango reviewed (commented):
A nice improvement. Few open comments, but otherwise LGTM

gautierg-st commented on drivers/adc/adc_stm32.c:
```diff
@@ -58,20 +58,26 @@ LOG_MODULE_REGISTER(adc_stm32);
+#elif defined(CONFIG_SOC_SERIES_STM32H7X)
+#if defined(ADC_VER_V5_V90)
+#define STM32H72X_ADC /* H72x and H73x */
+#elif defined(ADC_VER_V5_X)
+#define STM32H74X_ADC /* H74x and H75x */
```
I don't see how. ADC_VER_V5_X is common to all U5, and there's nothing that would group U5 and H74x without the other H7.

gautierg-st commented on dts/bindings/adc/st,stm32-adc.yaml:
```diff
@@ -104,5 +104,18 @@ properties:
+    type: int
+    required: true
+    enum:
+      - 0 # OVERSAMPLER_NONE
+      - 1 # OVERSAMPLER_MINIMAL
+      - 2 # OVERSAMPLER_EXTENDED
```
Will do. And will do the same for the sequencer while I'm at it

gautierg-st replied:
Done

gautierg-st committed:
dts: bindings: adc: stm32: use string enum instead of number enum

STM32 sequencer property and clock source were defined using
arbitrary numbers. Use string instead.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
dts: arm: st: change sequencer and clock source properties into string

In all STM32 dtsi and board dts, update the st,adc-sequencer and the
st,adc-clock-source properties so they are strings.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

fu dts: arm: st: use string instead of enum

gautierg-st committed:
drivers: adc: stm32: use a string for sequencer and clock source property

Now that st,adc-sequencer and st,adc_clock-source use a string, update the
ADC driver.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

fu drv adc update driver with string

gautierg-st commented on drivers/adc/adc_stm32.c:
```diff
@@ -793,10 +793,13 @@ static void dma_callback(const struct device *dev, void *user_data,
+		if (LL_ADC_IsActiveFlag_OVR(adc)) {
+			LL_ADC_ClearFlag_OVR(adc);
+			LOG_ERR("ADC overrun error occurred. Reduce clock source frequency, "
+				"increase prescaler value, or increase sampling times.");
```
Done

gautierg-st committed:
drivers: adc: stm32: listify some tables

Listify the content of the tables used for sequencer and oversampling.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

bjarki-andreasen commented on drivers/adc/adc_stm32.c:
```diff
@@ -734,8 +669,11 @@ static void adc_stm32_oversampling_ratioshift(ADC_TypeDef *adc, uint32_t ratio,
+static int adc_stm32_oversampling(const struct device *dev, uint8_t ratio)
 {
+	const struct adc_stm32_cfg *config = dev->config;
+	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
```
The typecast here is not needed, its already `ADC_TypeDef *`

bjarki-andreasen commented on include/zephyr/dt-bindings/adc/stm32_adc.h:
```diff
@@ -84,4 +84,15 @@
 #define FULLY_CONFIGURABLE	1
 /** @} */
 
+/**
+ * @name STM32 ADC oversampler type
+ * This value is to set <st,adc-oversampler>
+ * Refer to the RefMan to know which type corresponds to which ADC.
+ * @{
+ */
+#define OVERSAMPLER_NONE	0	/* No oversampler */
+#define OVERSAMPLER_MINIMAL	1	/* 8 oversampling values available */
+#define OVERSAMPLER_EXTENDED	2	/* 1024 oversampling values available */
+/** @} */
+
```
Is this used? it seems this PR decided on using strings for `OVERSAMPLER_` and `_CONFIGURABLE` enums

bjarki-andreasen replied:
looking a bit further, the defines are used in the driver, not in the devicetree, so I would move these inside the driver as they seemingly don't need to be exposed to the devicetree :)

gautierg-st replied:
You're right, they're not needed here anymore. Moved into the driver

gautierg-st committed:
drivers: adc: stm32: remove unused cast

config->base is already defined as ADC_TypeDef so no there is no need to
cast it as such. Remove all occurrences throughout the file.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: stm32: move define from dt-binding into driver

Now that clock source and sequencer are defined with strings in device
tree, move the old defines directly in the driver

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st commented:
Also applied string enum to st,adc-clock-source property that I missed earlier

bjarki-andreasen reviewed (approved):
Nice work!

benediktibk commented:
Could it be that there is an entry in the release notes/migration guide missing? Because changing the type of the properties to string is somewhat a breaking change. I just stumbled upon this. No big deal, but might be noteworthy in the documentation.

kartben commented:
> Could it be that there is an entry in the release notes/migration guide missing? Because changing the type of the properties to string is somewhat a breaking change. I just stumbled upon this. No big deal, but might be noteworthy in the documentation.

+1. @gautierg-st please follow up on adding this to the migration guide.

gautierg-st commented:
> > Could it be that there is an entry in the release notes/migration guide missing? Because changing the type of the properties to string is somewhat a breaking change. I just stumbled upon this. No big deal, but might be noteworthy in the documentation.
> 
> +1. @gautierg-st please follow up on adding this to the migration guide.

Done here #85643
